Shared/mods/deathmatch/logic/lua/CLuaFunctionParser.h

```diff
@@ -0,0 +1,443 @@
+
+                std::size_t i = -2;
+                auto        k = PopUnsafe<key_t>(L, i);
+                auto        v = PopUnsafe<value_t>(L, i);
+                map.emplace(std::move(k), std::move(v));
+                lua_pop(L, 1);            // drop value, keep key for lua_next
+            }
+            ++index;
+            return map;
+        }
+        else if constexpr (std::is_same_v<T, CLuaFunctionRef>)
+        {
+            return luaM_toref(L, index++);
+        }
+        else if constexpr (std::is_same_v<T, lua_State*>)
+            return L;
+        // variants can be used by any of the underlying types
+        // thus recursively use this function
+        else if constexpr (is_variant<T>::value)
+        {
+            int iMatch = TypeMatchVariant<T>(L, index);
+            return PopUnsafeVariant<T>(L, index, iMatch);
+        }
+
+        // Catch all for class pointer types, assume all classes are valid script entities
+        // and can be fetched from a userdata
+        else if constexpr (std::is_pointer_v<T> && std::is_class_v<std::remove_pointer_t<T>>)
+        {
+            bool  isLightUserData = lua_type(L, index) == LUA_TLIGHTUSERDATA;
+            void* pValue = lua::PopPrimitive<void*>(L, index);
+            using class_t = std::remove_pointer_t<T>;
+            auto result =
+                isLightUserData ? UserDataCast<class_t>((class_t*)0, pValue, L) : UserDataCast<class_t>((class_t*)0, *reinterpret_cast<void**>(pValue), L);
+            if (result == nullptr)
+            {
+                SString strReceived = isLightUserData ? GetUserDataClassName(pValue, L) : GetUserDataClassName(*(void**)pValue, L);
+                SString strExpected = GetClassTypeName((T)0);
+                SString strMessage("Expected %s at argument %d, got %s", strExpected.c_str(), index, strReceived.c_str());
+                strError = strMessage;
+                return nullptr;
+            }
+            return static_cast<T>(result);
+        }
+    }
+};
+
+template <bool, auto, auto*>
+struct CLuaFunctionParser
+{
+};
+
+template <bool ErrorOnFailure, auto ReturnOnFailure, typename Ret, typename... Args, auto (*Func)(Args...)->Ret>
+struct CLuaFunctionParser<ErrorOnFailure, ReturnOnFailure, Func> : CLuaFunctionParserBase
+{
+    template <typename... Params>
+    inline auto Call(lua_State* L, Params&&... ps)
+    {
+        if (strError.length() != 0)
+        {
+            return -1;
+        }
+        if constexpr (sizeof...(Params) == sizeof...(Args))
+        {
+            if constexpr (std::is_same_v<Ret, void>)
+            {
+                Func(std::forward<Params>(ps)...);
+                return 0;
+            }
+            else
+            {
+                return lua::Push(L, Func(std::forward<Params>(ps)...));
+            }
+        }
+        else
+        {
+            return Call(L, ps..., Pop<typename nth_element_impl<sizeof...(Params), Args...>::type>(L, iIndex));
```

Comment:
What is this for?

Reply:
Call recursively calls itself until it has reached the desired number of arguments. In each step we pop the next parameter from the Lua stack and append it to the list of arguments.

Reply:
And then Push the return value?

Reply:
Yup. So if we're trying to call `int foo(std::string, float)` we get the following calls:
`Call(L);`
`Call(L, str);` (with str being the string returned from Pop)
`Call(L, str, flt)` (flt being the float from Pop)
`Push(Func(str, flt));`